java/ql/lib/semmle/code/java/security/AndroidIntentRedirection.qll

```diff
@@ -0,0 +1,81 @@
+    row =
+      [
+        "android.app;Activity;true;bindService;;;Argument[0];intent-start",
+        "android.app;Activity;true;bindServiceAsUser;;;Argument[0];intent-start",
+        "android.app;Activity;true;startActivityAsCaller;;;Argument[0];intent-start",
+        "android.app;Activity;true;startActivityForResult;(Intent,int);;Argument[0];intent-start",
+        "android.app;Activity;true;startActivityForResult;(Intent,int,Bundle);;Argument[0];intent-start",
+        "android.app;Activity;true;startActivityForResult;(String,Intent,int,Bundle);;Argument[1];intent-start",
+        "android.app;Activity;true;startActivityForResultAsUser;;;Argument[0];intent-start",
+        "android.content;Context;true;startActivities;;;Argument[0];intent-start",
+        "android.content;Context;true;startActivity;;;Argument[0];intent-start",
+        "android.content;Context;true;startActivityAsUser;;;Argument[0];intent-start",
+        "android.content;Context;true;startActivityFromChild;;;Argument[1];intent-start",
+        "android.content;Context;true;startActivityFromFragment;;;Argument[1];intent-start",
+        "android.content;Context;true;startActivityIfNeeded;;;Argument[0];intent-start",
+        "android.content;Context;true;startForegroundService;;;Argument[0];intent-start",
+        "android.content;Context;true;startService;;;Argument[0];intent-start",
+        "android.content;Context;true;startServiceAsUser;;;Argument[0];intent-start",
+        "android.content;Context;true;sendBroadcast;;;Argument[0];intent-start",
+        "android.content;Context;true;sendBroadcastAsUser;;;Argument[0];intent-start",
+        "android.content;Context;true;sendBroadcastWithMultiplePermissions;;;Argument[0];intent-start",
+        "android.content;Context;true;sendStickyBroadcast;;;Argument[0];intent-start",
+        "android.content;Context;true;sendStickyBroadcastAsUser;;;Argument[0];intent-start",
+        "android.content;Context;true;sendStickyOrderedBroadcast;;;Argument[0];intent-start",
+        "android.content;Context;true;sendStickyOrderedBroadcastAsUser;;;Argument[0];intent-start"
+      ]
+  }
+}
+
+/** Default sink for Intent redirection vulnerabilities. */
+private class DefaultIntentRedirectionSink extends IntentRedirectionSink {
+  DefaultIntentRedirectionSink() { sinkNode(this, "intent-start") }
+}
+
+/**
+ * A default sanitizer for nodes dominated by calls to `ComponentName.getPackageName`
+ * or `ComponentName.getClassName`. These are used to check whether the origin or destination
+ * components are trusted.
+ */
+private class DefaultIntentRedirectionSanitizer extends IntentRedirectionSanitizer {
+  DefaultIntentRedirectionSanitizer() {
+    exists(MethodAccess ma, Method m |
+      ma.getMethod() = m and
+      m.getDeclaringType() instanceof TypeComponentName and
+      m.hasName(["getPackageName", "getClassName"]) and
+      ma.getBasicBlock().(ConditionBlock).controls(this.asExpr().getBasicBlock(), true)
+    )
```

Comment:
This is a little too ad-hoc, try this instead:
```suggestion
    exists(MethodAccess ma, Method m, Guard g, boolean branch |
      ma.getMethod() = m and
      m.getDeclaringType() instanceof TypeComponentName and
      m.hasName(["getPackageName", "getClassName"]) and
      g.isEquality(ma, _, branch) and
      g.controls(this.asExpr().getBasicBlock(), branch)
    )
```

Reply:
That seems more sensible indeed. After your suggestion, I thought maybe a `BarrierGuard` would be even more appropriate for this, but in the end I think your approach is simpler as-is. Thank you!